### PR TITLE
fix DragControl not work on mobile

### PR DESCRIPTION
**Pull request by chunnallu**
I have fixed DragControl to work on mobile , and verified this not break the exist example --
webgl_interactive_draggablecubes.html

### CODE REVIEW

File: examples/js/controls/DragControls.js [hunk 1]
```diff
@@ -39,6 +42,10 @@ THREE.DragControls = function ( _objects, _camera, _domElement ) {
 		_domElement.removeEventListener( 'mousemove', onDocumentMouseMove, false );
 		_domElement.removeEventListener( 'mousedown', onDocumentMouseDown, false );
 		_domElement.removeEventListener( 'mouseup', onDocumentMouseUp, false );
+                _domElement.removeEventListener( 'touchmove', onDocumentMouseMove, false );
+                _domElement.removeEventListener( 'touchstart', onDocumentMouseDown, false );
+                _domElement.removeEventListener( 'touchend', onDocumentMouseUp, false );
+
```
Comment (Mugen87) [inline examples/js/controls/DragControls.js]: Some code style feedback: Indentation should be tabs not spaced.

File: examples/js/controls/DragControls.js [hunk 2]
```diff
@@ -51,11 +58,12 @@ THREE.DragControls = function ( _objects, _camera, _domElement ) {
...
 
 		event.preventDefault();
+                pointer = event.changedTouches ? event.changedTouches[0] : event;
 
 		var rect = _domElement.getBoundingClientRect();
 
-		_mouse.x = ( (event.clientX - rect.left) / rect.width ) * 2 - 1;
-		_mouse.y = - ( (event.clientY - rect.top) / rect.height ) * 2 + 1;
+		_mouse.x = ( (pointer.clientX - rect.left) / rect.width ) * 2 - 1;
+		_mouse.y = - ( (pointer.clientY - rect.top) / rect.height ) * 2 + 1;
```
Comment (Mugen87) [inline examples/js/controls/DragControls.js]: Here to...
```javascript
_mouse.x = ( ( pointer.clientX - rect.left ) / rect.width ) * 2 - 1;
```
Reference https://github.com/mrdoob/three.js/wiki/Mr.doob%27s-Code-Style%E2%84%A2

File: examples/js/controls/DragControls.js [hunk 3]
```diff
@@ -51,11 +58,12 @@ THREE.DragControls = function ( _objects, _camera, _domElement ) {
 	function onDocumentMouseMove( event ) {
 
 		event.preventDefault();
+                pointer = event.changedTouches ? event.changedTouches[0] : event;
```
Comment (Mugen87) [inline examples/js/controls/DragControls.js]: Please insert spaces between parenthesis like so:
```javascript
event.changedTouches[ 0 ]
```
Besides, it looks like `pointer` is an undeclared variable.

File: examples/js/controls/DragControls.js [hunk 4]
```diff
@@ -66,9 +74,9 @@ THREE.DragControls = function ( _objects, _camera, _domElement ) {
 				_selected.position.copy( _intersection.sub( _offset ) );
 
 			}
-
```
Comment (Mugen87) [inline examples/js/controls/DragControls.js]: Not necessary to delete this line...

File: examples/js/controls/DragControls.js [hunk 5]
```diff
@@ -66,9 +74,9 @@ THREE.DragControls = function ( _objects, _camera, _domElement ) {
...
 			scope.dispatchEvent( { type: 'drag', object: _selected } );
 
+
```
Comment (Mugen87) [inline examples/js/controls/DragControls.js]: This insertion is not needed...

File: examples/js/controls/DragControls.js [hunk 6]
```diff
@@ -119,6 +133,8 @@ THREE.DragControls = function ( _objects, _camera, _domElement ) {
 
 			_selected = intersects[ 0 ].object;
 
+                        _plane.setFromNormalAndCoplanarPoint( _camera.getWorldDirection( _plane.normal ), _selected.position );
```
Comment (Mugen87) [inline examples/js/controls/DragControls.js]: Why have added this line? Isn't `_plane` already calculated?

**Review by Mugen87 — CHANGES_REQUESTED** [review]

**Comment by Mugen87** [comment]
BTW: I'm not sure it's a good idea to use the mouse event handlers for `touchmove`, `touchstart` and `touchend`. Touch interaction proceeds differently than with the mouse. So for example you perform operations on `touchmove` that are actually not required, right?

**Comment by Mugen87** [comment]
Have you considered to create new event listeners for `touchmove`, `touchstart` and `touchend`? Maybe this makes the code simpler to read...

File: examples/js/controls/DragControls.js (hunk 6 above)
Comment (chunnallu) [inline examples/js/controls/DragControls.js]: That is right !

**Comment by chunnallu** [comment]
thanks @Mugen87 , I will add some  changes to those code

**Comment by chunnallu** [comment]
to make some change

**Comment by chunnallu** [comment]
still some code style issue

**Comment by chunnallu** [comment]
have fix the code style issue , but have not `create new event listeners for touchmove, touchstart and touchend instead of reusing the mouse event handlers ` , becasuse touch event handler are very similar to mouse event handler , create new event listeners may add some redundancy code @Mugen87

**Comment by Mugen87** [comment]
@mrdoob I'm not sure if `THREE.DragControls` needs some refactoring now. As @chunnallu said, new event listeners for touch events may blow up the code. On the other side, it would be easier to separate the logic. Redundant code could be implemented in new shared methods.

For example, `THREE.DragControls` fires `hoveron` and `hoveroff` events which makes sense for mouse controls but not for touch devices. I think it would be nice if touch related event listeners don't try to fire these events.

**Comment by mrdoob** [comment]
How about switching to pointer events?

**Comment by chunnallu** [comment]
> How about switching to pointer events?

Is it mean that refactor event type `hoveron` to  event type `pointer` ? like below :

from :
```
scope.dispatchEvent( { type: 'hoveron', object: object } );
```
to : 
```
scope.dispatchEvent( { type: 'pointer', object: object } );
```
@mrdoob @Mugen87

**Comment by Mugen87** [comment]
I think @mrdoob means:

- `pointerdown` replaces  `touchstart` and `mousedown`
- `pointerup` replaces `touchend` and `mouseup`
- `pointermove` replaces `touchmove` and `mousemove`

@chunnallu Do you think you can change so `THREE.DragControl` uses pointer events.

As i mentioned before, i still think it makes no sense that touch devices fire hover events. We maybe should query [pointerEvent.pointerType](https://developer.mozilla.org/en-US/docs/Web/API/PointerEvent/pointerType) and only fire hover events for the value `mouse`.

**Comment by chunnallu** [comment]
maybe i can ! In fact i am a fresh coder , i do this maybe slow with many error，haha ! so  i think the best way is that you or someone other in rich experience to change the code @Mugen87

**Comment by Mugen87** [comment]
@chunnallu Give it a try! This is the best way to learn programming :blush:

**Comment by chunnallu** [comment]
thanks ! haha , can't not aggre more!

**Comment by chunnallu** [comment]
new commit : 

- separate the logic of touch event and mouse event , create new event listeners for touchend

- add `mouseleave` event 

- rename `onDocumentMouseUp` handler to `onDocumentMouseCancel` , so it can use to handle `mouseup` and 'mouseleave` event

beside , use pointer events seem to be a little unclear , so  i still to separate the event handles. :stuck_out_tongue_winking_eye:

**Comment by Mugen87** [comment]
I've realized right now that the browser support for Pointer events is not good enough, see http://caniuse.com/#feat=pointer

So for the moment it's maybe better to stay at `mouse` and `touch` events even if this means some redundant code.

**Review by Mugen87 — APPROVED** [review]

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**